src/ray/gcs/gcs_client/gcs_client.cc

```diff
@@ -88,42 +88,49 @@ GcsClient::GcsClient(
 
 Status GcsClient::Connect(instrumented_io_context &io_service) {
   RAY_CHECK(!is_connected_);
-
-  if (options_.server_ip_.empty()) {
-    RAY_LOG(ERROR) << "Failed to connect, gcs service address is empty.";
+  if (options_.redis_ip_.empty() && options_.gcs_address_.empty()) {
+    RAY_LOG(ERROR) << "Failed to connect, server ip and gcs address both are empty.";
     return Status::Invalid("gcs service address is invalid!");
   }
 
-  // Connect to redis.
-  // We don't access redis shardings in GCS client, so we set `enable_sharding_conn` to
-  // false.
-  RedisClientOptions redis_client_options(
-      options_.server_ip_, options_.server_port_, options_.password_,
-      /*enable_sharding_conn=*/false, options_.enable_sync_conn_,
-      options_.enable_async_conn_, options_.enable_subscribe_conn_);
-  redis_client_ = std::make_shared<RedisClient>(redis_client_options);
-  RAY_CHECK_OK(redis_client_->Connect(io_service));
-
-  // Get gcs service address.
-  if (get_server_address_func_) {
-    get_server_address_func_(&current_gcs_server_address_);
-    int i = 0;
-    while (current_gcs_server_address_.first.empty() &&
-           i < RayConfig::instance().gcs_service_connect_retries()) {
-      std::this_thread::sleep_for(std::chrono::milliseconds(
-          RayConfig::instance().internal_gcs_service_connect_wait_milliseconds()));
-      get_server_address_func_(&current_gcs_server_address_);
-      i++;
-    }
+  if (options_.gcs_address_.empty()) {
+    // Connect to redis.
+    // We don't access redis shardings in GCS client, so we set `enable_sharding_conn`
+    // to false.
+    RedisClientOptions redis_client_options(
+        options_.redis_ip_, options_.redis_port_, options_.password_,
+        /*enable_sharding_conn=*/false, options_.enable_sync_conn_,
+        options_.enable_async_conn_, options_.enable_subscribe_conn_);
+    redis_client_ = std::make_shared<RedisClient>(redis_client_options);
+    RAY_CHECK_OK(redis_client_->Connect(io_service));
   } else {
-    get_server_address_func_ = [this](std::pair<std::string, int> *address) {
-      return GetGcsServerAddressFromRedis(
-          redis_client_->GetPrimaryContext()->sync_context(), address);
-    };
-    RAY_CHECK(GetGcsServerAddressFromRedis(
-        redis_client_->GetPrimaryContext()->sync_context(), &current_gcs_server_address_,
-        RayConfig::instance().gcs_service_connect_retries()))
-        << "Failed to get gcs server address when init gcs client.";
+    RAY_CHECK(::RayConfig::instance().gcs_grpc_based_pubsub())
+        << "If using gcs_address to start client, gRPC based pubsub has to be enabled";
+  }
+
+  // Setup gcs server address fetcher
+  if (get_server_address_func_ == nullptr) {
```

Comment:
How about this?
```
if (get_server_address_func_ != nullptr) {
  // call it once to get gcs address and port
} else if (!options_.gcs_address_.empty()) {
  // set gcs address and port from options_
} else {
  // resolve gcs address and port from redis_client_
}
```
Then we would be able to delete the redis branch after the migration. Also only the redis branch needs retries.

Reply:
The flow is like this:
1. if get_server_address_func_ is set, use this one directly.
2. if get_server_address_func_ is not set, set it
  1. use gcs address
  2. or read from redis


get_server_address_func_ will be used in other places too.

Reply:
I understand the logic in the PR. I just think that
1. Generating `get_server_address_func_` from GCS address in `GcsClientOptions`, then getting gcs address from the function seems unnecessary.
1. Retrying `get_server_address_func_` is only useful when reading GCS address from Redis.
So I'm hoping we only use `get_server_address_func_` when it is set, which does not need to be called elsewhere. What are the places you see where `get_server_address_func_` will be useful?

Reply:
@mwtian `get_server_address_func_` is passed from the constructor, not always generated here. 

Another thing here is that, it doesn't hurt to use get_server_address_func_ to read the address. If it's redis, read from redis, if it's fixed, return the fixed one. We don't need extra if-else for later.

Reply:
I have checked the `get_server_address_func_` passed in via constructor, and it was just sending GCS address values. This is the reason retry was not needed before. My proposal does not update `get_server_address_func_`. I still think it is confusing to retry on a function returning in memory values. But if you insist we can keep the logic.

Reply:
@mwtian  I'll keep this variable and also the constructor. It looks like we have logic to get gcs address from raylet [code](https://sourcegraph.com/github.com/ray-project/ray@30d3115c4585df3b5cead882b69c7a80c8f023b6/-/blob/src/ray/core_worker/gcs_server_address_updater.cc?L20)

We can get rid of this when we do clean up.